dashboard/app/mailers/pd/workshop_mailer.rb

```diff
@@ -87,22 +87,34 @@ def teacher_enrollment_reminder(enrollment)
       reply_to: email_address(@workshop.organizer.name, @workshop.organizer.email)
   end
 
-  # TODO: Make sure every teacher has an enrollment
   def exit_survey(workshop, teacher, enrollment)
     @workshop = workshop
     @teacher = teacher
     @is_first_workshop = Pd::Workshop.attended_by(teacher).in_state(Pd::Workshop::STATE_ENDED).count == 1
 
-    @survey_url = CDO.code_org_url "/pd-workshop-survey/#{enrollment.code}"
+    @survey_url = CDO.code_org_url "/pd-workshop-survey/#{enrollment.code}", 'https:'
 
     mail content_type: 'text/html',
       from: from_hadi,
       subject: 'How was your Code.org workshop?',
       to: email_address(@teacher.name, @teacher.email)
+
+    attachments['certificate.jpg'] = generate_csf_certificate if @workshop.course == Pd::Workshop::COURSE_CSF
   end
 
   private
 
+  def generate_csf_certificate
+    image = create_certificate_image2(
+      dashboard_dir('app', 'assets', 'images', 'pd_workshop_certificate_csf.png'),
+      @teacher.name || @teacher.email,
+      y: 444,
+      height: 100,
+    )
+    image.format = 'jpg'
+    Base64.encode64(image.to_blob)
```

Comment:
The teacher-supplied name is written as-is onto the certificate image in `create_certificate_image2`. Then `image.to_blob` is dumping the image binary, and `encode64` is encoding the resulting image to be used as an email attachment. Since it's sent as a MIME attachment and not used as part of a url, it doesn't need to be url-safe. Also note this was basically copied from the old [certificate email](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/emails/professional-development-workshop-section-receipt.md).